source/mir/base64.d

```diff
@@ -0,0 +1,425 @@
+const(char)[] encodeBase64(char PlusChar = '+', char SlashChar = '/')(scope const(ubyte)[] buf) @safe pure
+{
+    import mir.appender : scopedBuffer;
+    // XXX: is a stringBuf more appropriate here?
+    auto app = scopedBuffer!char;
+    encodeBase64!(PlusChar, SlashChar)(buf, app);
+    return app.data.dup;
+}
+
+/++
+Encode a ubyte array as Base64, placing the result onto an Appender.
++/
+void encodeBase64(char PlusChar = '+', char SlashChar = '/', Appender)(scope const(ubyte)[] input,
+                                                                       scope return ref Appender appender) @safe pure
+{
+    import mir.ndslice.topology : bytegroup, map;
+    import core.bitop : bswap;
+    // Slice our input array so that n % 3 == 0 (we have a multiple of 3) 
+    // If we have less then 3, then this is effectively a no-op (will result in a 0-length slice)
+    char[4] encodedByteGroup;
+    const(ubyte)[] window = input[0 .. input.length - (input.length % 3)];
+    foreach(group; window.bytegroup!(3, uint).map!bswap) {
+        const(ubyte) a = (group >> 26) & 0x3f;
+        const(ubyte) b = (group >> 20) & 0x3f;
+        const(ubyte) c = (group >> 14) & 0x3f;
+        const(ubyte) d = (group >> 8) & 0x3f;
+
+        encodedByteGroup[0] = a.lookup_encoding!(PlusChar, SlashChar);
+        encodedByteGroup[1] = b.lookup_encoding!(PlusChar, SlashChar);
+        encodedByteGroup[2] = c.lookup_encoding!(PlusChar, SlashChar);
+        encodedByteGroup[3] = d.lookup_encoding!(PlusChar, SlashChar);
+        appender.put(encodedByteGroup[]);
+    }
+
+    // If it's a clean multiple of 3, then it requires no padding.
+    // If not, then we need to add padding.
+    if (input.length % 3 != 0)
+    {
+        window = input[window.length .. input.length];
+
+        uint group = (window[0] << 24);
+
+        if (window.length == 1) {
+            const(ubyte) a = (group >> 26) & 0x3f;
+            const(ubyte) b = (group >> 20) & 0x3f;
+            encodedByteGroup[0] = a.lookup_encoding!(PlusChar, SlashChar);
+            encodedByteGroup[1] = b.lookup_encoding!(PlusChar, SlashChar);
+            encodedByteGroup[2] = '=';
+            encodedByteGroup[3] = '=';
+            appender.put(encodedByteGroup[]);
+        }
+        else {
+            // Just in case math fails or something
+            assert(window.length == 2);
+
+            group |= (window[1] << 16);
+            const(ubyte) a = (group >> 26) & 0x3f;
+            const(ubyte) b = (group >> 20) & 0x3f;
+            const(ubyte) c = (group >> 14) & 0x3f;
+            encodedByteGroup[0] = a.lookup_encoding!(PlusChar, SlashChar);
+            encodedByteGroup[1] = b.lookup_encoding!(PlusChar, SlashChar);
+            encodedByteGroup[2] = c.lookup_encoding!(PlusChar, SlashChar);
+            encodedByteGroup[3] = '=';
+            appender.put(encodedByteGroup[]);
+        }
+    }
+}
+
+/// Test encoding of data which has a length that can be cleanly
+/// encoded.
+unittest
```

Comment:
Tests should be marked `@safe pure` and at least one `@safe pure @nogv`.
Both for encoding and decoding.